lib/charms/observability_libs/v1/kubernetes_compute_resources_patch.py

```
@@ -0,0 +1,631 @@
+
+    def get_templated(self) -> Optional[ResourceRequirements]:
+        """Returns the resource limits specified in the StatefulSet template."""
+        statefulset = self.client.get(
+            StatefulSet, name=self.statefulset_name, namespace=self.namespace
+        )
+        podspec_tpl = self._get_container(
+            self.container_name,
+            statefulset.spec.template.spec.containers,  # type: ignore[attr-defined]
+        )
+        return podspec_tpl.resources
+
+    def get_actual(self, pod_name: str) -> Optional[ResourceRequirements]:
+        """Return the resource limits that are in effect for the container in the given pod."""
+        pod = self.client.get(Pod, name=pod_name, namespace=self.namespace)
+        podspec = self._get_container(
+            self.container_name, pod.spec.containers  # type: ignore[attr-defined]
+        )
+        return podspec.resources
+
+    def is_ready(self, pod_name, resource_reqs: ResourceRequirements):
+        """Reports if the resource patch has been applied and is in effect.
+
+        Returns:
+            bool: A boolean indicating if the service patch has been applied and is in effect.
+        """
+        logger.info(
+            "reqs=%s, templated=%s, actual=%s",
+            resource_reqs,
+            self.get_templated(),
+            self.get_actual(pod_name),
+        )
+        return self.is_patched(resource_reqs) and equals_canonically(  # pyright: ignore
+            resource_reqs, self.get_actual(pod_name)  # pyright: ignore
+        )
+
+    def apply(self, resource_reqs: ResourceRequirements) -> None:
+        """Patch the Kubernetes resources created by Juju to limit cpu or mem."""
+        # Need to ignore invalid input, otherwise the StatefulSet gives "FailedCreate" and the
+        # charm would be stuck in unknown/lost.
+        if self.is_patched(resource_reqs):
+            return
```

I'd give the caller some indication of whether we did something or not. 
Either add a bool return value (True -> patch applied, we did do something, False -> no-op)
Or we could raise a custom exception so the user knows they should be checking if the patch is applied already.

---

the caller will get notified by an exception if things go sideways.
the caller interprets a "no exceptions fired" as an "everything is good".
its an internal call that will be handled by the caller function to nicely inform the user about it.

---

Added a debug log if its already patched